### PR TITLE
fix internal::has_begin type trait used in FEEvaluation

### PR DESCRIPTION
**Pull request by davydden**
minor fix and follow up to https://github.com/dealii/dealii/pull/7744 . The original implementation would only work if `T::const_iterator` exist, which defeats the point of the type trait.

### CODE REVIEW

File: include/deal.II/matrix_free/fe_evaluation.h [hunk 1]
```diff
@@ -3466,8 +3466,7 @@ namespace internal
 
   public:
     static constexpr bool value =
-      std::is_same<typename T::const_iterator,
-                   decltype(detect(std::declval<T>()))>::value;
+      !std::is_same<void, decltype(detect(std::declval<T>()))>::value;
```
Comment (masterleinad) [inline include/deal.II/matrix_free/fe_evaluation.h]: Can you update the comments in line 3404 and 3455 such that they show the return type we test for (first case) or don't (second case)?

Comment (davydden) [reply include/deal.II/matrix_free/fe_evaluation.h]: but those comments are NOT for this detector.

Comment (davydden) [reply include/deal.II/matrix_free/fe_evaluation.h]: at least `3404`. 

`3455` still applies as is. We test if  there `something ... T::begin()` where `something != void`.

Comment (davydden) [reply include/deal.II/matrix_free/fe_evaluation.h]: maybe I misunderstand what you would like to have in comments...

Comment (masterleinad) [reply include/deal.II/matrix_free/fe_evaluation.h]: Yes, line 3404 is not related to the pull request here, but I was too late to comment on the original pull request.
There we are saying 
```
 // ... T::local_element() const
```
but we are testing that the return type is `typename T::value_type`. What I wanted is to either be specific:
```
 // typename T::value_type T::local_element() const
```
or also test for any non-`void` return type there as well.

In line 3455, we are saying
```
 // T::const_iterator T::begin() const
```
but test for any non-`void` return type. In view of the style in line 3404, I expected

```
 // ... T::begin() const
```

Comment (davydden) [reply include/deal.II/matrix_free/fe_evaluation.h]: ah, ok, i will adjust the comments a bit later today

**Comment by masterleinad** [comment]
The test is failing for `clang-4.0.1`: https://cdash.kyomu.43-1.org/testDetails.php?test=54645309&build=8122

**Comment by davydden** [comment]
> The test is failing for `clang-4.0.1`:

I saw, I have no idea what is going on here... we do include a header so the functions should be there and defined.

**Comment by masterleinad** [comment]
The strategy that works with all the compilers for `static const(expr)` member variabels that are ODR-used is (compare `VectorizedArray::n_elements`):
- use `static const` and initialize in class
- provide an external definition

**Comment by davydden** [comment]
@masterleinad I addressed both issue, I believe.

**Comment by masterleinad** [comment]
Let's try that.

**Comment by davydden** [comment]
> Let's try that.

I think you still need to approve the PR before we can merge even if testers are happy ;-)

**Pull request merged**